### PR TITLE
#167313411 Users should be able to tag their articles

### PR DESCRIPTION
**Pull request by Anguandia**
#### What does this PR do?

  Add the 'tag articles feature'

#### Description of Task to be completed?

  - Implement article Tagging
  - Add article search by tag
  - Add orpahn and article tag modification
  - Add tag deletion
  - Add creation of individual tags
  - Add fetching all tags of a given article
  - Add deletion of a specific tag of a specific article

#### How should this be manually tested?

  - Tag creation:
    post/baseUrl/api/v1/tags request with a body {name: Your tag name}
  - Tagging an article:
    post/baseUrl/articles/:articleId/:your-tag-name
  - Updating a tag:
    patch/base-url/tags/:old tag name with a body {update: new tag name}
  - Updating a specific tag of an article:
    patch/articles/:articleId/:tagName with a body {name: tagName}
  - Getting all articles with a given tag:
    get/tags/:tagName/articles
  - Getting all tags of a given article:
    get/articles/:articleId/tags
  - Getting all tags:
    get/tags
  - Deleting a tag:
    delete/tags/:tagName
  - Deleting a specific tag of a specific article
    delete/articles/:articleId/:tagName

#### What are the relevant pivotal tracker stories?

  [#167313411](https://www.pivotaltracker.com/story/show/167313411)

### CODE REVIEW

**Review by Mnickii — CHANGES_REQUESTED** [review]
Let's increase the coverage on this one, and edit the PR title according to the standards we follow.

File: src/controllers/tag.js [hunk 1]
```diff
@@ -0,0 +1,196 @@
+/* eslint-disable require-jsdoc */
```
Comment (nkalyesubula) [inline src/controllers/tag.js:1]: Could use remove this

File: src/controllers/tag.js [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    tags.forEach((name) => {
+      Tag.findOrCreate({
+        where: { name }
+      }).then(tag => article.addTag(tag[0]));
+    });
+    return addTags([added, exists, article]).send(res);
+  }
+
+  static async editTag(req, res) {
+    // changes the name of a given tag if new name still unique or merges tags
```
Comment (nkalyesubula) [inline src/controllers/tag.js:54]: You could use doc-strings for a multiple line comment

File: src/controllers/tag.js [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    }).then((article) => {
+      if (article.tags.length === 0) {
+        return res.status(200).json({ message: 'article not tagged' });
+      }
+      res.status(200).json({ tags: article.tags.map(tag => tag.name) });
+    }).catch(() => internalError().send(res));
+  }
+
+  static async deleteArticleTag(req, res) {
+    // deletes a given tag of a given article technically the tag and article
```
Comment (nkalyesubula) [inline src/controllers/tag.js:151]: same here

**Review by nkalyesubula — CHANGES_REQUESTED** [review]
Good work Mike. Could you please add screenshots in your PR, and  also address other comments have left

**Comment by Mnickii** [comment]
@Anguandia in the meantime you could attend to the reviews

File: src/controllers/tag.js (hunk 2 above)
Comment (Anguandia) [reply src/controllers/tag.js:54]: Thanks for the observations @nkalyesubula, implemented

**Comment by Anguandia** [comment]
> @Anguandia in the meantime you could attend to the reviews

Thanks, tests improved and  comments removed.
working on converting then .... catch to try ... catch to eliminate catch(500) impact on coverage.
Thanks a great deal

**Review by habinezadalvan — COMMENTED** [review]
Please rebase

**Pull request closed without merging**